File: src/algebra2_rational_expressions_easy.py
"""
algebra2 - rational_expressions (easy)
Generated: 2026-02-22T04:23:13.838829
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Type 1: Simplify a simple rational expression (single fraction)
        # ELO 1000-1100: Basic simplification with common factors
        factor1 = nonzero(-5, 5)
        factor2 = nonzero(-5, 5)
        common = nonzero(2, 4)
        
        numerator = common * factor1
        denominator = common * factor2
        
        expr = numerator * x / (denominator * x)
        ans = simplify(expr)
        
        return problem(
            question=f"Simplify: $\\frac{{{latex(numerator * x)}}}{{{latex(denominator * x)}}}$",
            answer=ans,
            difficulty=(1000, 1100),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"Factor out common factor from numerator and denominator",
                f"$\\frac{{{latex(numerator * x)}}}{{{latex(denominator * x)}}} = \\frac{{{common} \\cdot {factor1} \\cdot x}}{{{common} \\cdot {factor2} \\cdot x}}$",
                f"Cancel common factors of ${common}$ and $x$",
                f"${latex(ans)}$"
            ),
        )
    
    elif problem_type == 2:
        # Type 2: Add/subtract rational expressions with same denominator
        # ELO 1100-1200: Same denominator, combine numerators
        denom = nonzero(2, 6)
        num1_coeff = nonzero(-5, 5)
        num2_coeff = nonzero(-5, 5)
        
        expr1 = num1_coeff * x / denom
        expr2 = num2_coeff * x / denom
        
        operation = choice(['+', '-'])
        if operation == '+':
            expr = expr1 + expr2
            ans = simplify(expr)
            op_symbol = "+"
        else:
            expr = expr1 - expr2
            ans = simplify(expr)
            op_symbol = "-"
        
        return problem(
            question=f"Simplify: $\\frac{{{latex(num1_coeff * x)}}}{{{denom}}} {op_symbol} \\frac{{{latex(num2_coeff * x)}}}{{{denom}}}$",
            answer=ans,
            difficulty=(1100, 1200),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"Since denominators are the same, combine numerators",
                f"$\\frac{{{latex(num1_coeff * x)} {op_symbol} {latex(num2_coeff * x)}}}{{{denom}}}$",
                f"Simplify the numerator: ${latex(ans)}$"
            ),
        )
    
    elif problem_type == 3:
        # Type 3: Multiply two simple rational expressions
        # ELO 1200-1300: Multiply and simplify
        num1 = nonzero(2, 5)
        denom1 = nonzero(2, 5)
        num2 = nonzero(2, 5)
        denom2 = nonzero(2, 5)
        
        expr = (num1 * x / denom1) * (num2 / (denom2 * x))
        ans = simplify(expr)
        
        return problem(
            question=f"Multiply and simplify: $\\frac{{{latex(num1 * x)}}}{{{denom1}}} \\cdot \\frac{{{num2}}}{{{latex(denom2 * x)}}}$",
            answer=ans,
            difficulty=(1200, 1300),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"Multiply numerators and denominators",
                f"$\\frac{{{latex(num1 * x)} \\cdot {num2}}}{{{denom1} \\cdot {latex(denom2 * x)}}}$",
                f"$\\frac{{{latex(num1 * num2 * x)}}}{{{latex(denom1 * denom2 * x)}}}$",
                f"Cancel common factor $x$",
                f"${latex(ans)}$"
            ),
        )
    
    else:
        # Type 4: Divide two simple rational expressions
        # ELO 1200-1300: Divide (multiply by reciprocal)
        num1 = nonzero(2, 6)
        denom1 = nonzero(2, 6)
        num2 = nonzero(2, 6)
        denom2 = nonzero(2, 6)
        
        expr = (num1 * x / denom1) / (num2 * x / denom2)
        ans = simplify(expr)
        
        return problem(
            question=f"Divide and simplify: $\\frac{{{latex(num1 * x)}}}{{{denom1}}} \\div \\frac{{{latex(num2 * x)}}}{{{denom2}}}$",
            answer=ans,
            difficulty=(1200, 1300),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"Multiply by the reciprocal",
                f"$\\frac{{{latex(num1 * x)}}}{{{denom1}}} \\cdot \\frac{{{denom2}}}{{{latex(num2 * x)}}}$",
                f"$\\frac{{{latex(num1 * x)} \\cdot {denom2}}}{{{denom1} \\cdot {latex(num2 * x)}}}$",
                f"Cancel common factor $x$",
                f"${latex(ans)}$"
            ),
        )

emit(generate())

File: src/algebra2_rational_expressions_easy.jl
# algebra2 - rational_expressions (easy)
# Generated: 2026-03-08T20:15:40.626009

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra2/rational_expressions")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Simplify a rational expression with common factors
        # ELO: 700-900
        root = nonzero(-8, 8)
        a = nonzero(-5, 5)
        b = randint(-12, 12)
        
        numerator = expand(a * (x - root))
        denominator = expand((x - root) * (x - b))
        
        simplified_num = a
        simplified_den = x - b
        
        problem(
            question="Simplify: \$\\frac{$(tex(numerator))}{$(tex(denominator))}\$",
            answer="$(tex(simplified_num))/$(tex(simplified_den))",
            difficulty=(700, 900),
            solution=steps(
                sol("Given", numerator / denominator),
                "Factor numerator: \$$(tex(a*(x - root)))\$",
                "Factor denominator: \$$(tex((x - root)*(x - b)))\$",
                "Cancel common factor \$(x - $(root))\$",
                sol("Answer", simplified_num / simplified_den)
            ),
            time=90
        )
        
    elseif problem_type == 2
        # Add two rational expressions with same denominator
        # ELO: 750-950
        den_lin = rand_linear(x; a=(-6,6), b=(-10,10))
        num1 = rand_linear(x; a=(-5,5), b=(-8,8))
        num2 = rand_linear(x; a=(-5,5), b=(-8,8))
        
        sum_num = simplify(num1.expr + num2.expr)
        
        problem(
            question="Add: \$\\frac{$(tex(num1.expr))}{$(tex(den_lin.expr))} + \\frac{$(tex(num2.expr))}{$(tex(den_lin.expr))}\$",
            answer="$(tex(sum_num))/$(tex(den_lin.expr))",
            difficulty=(750, 950),
            solution=steps(
                "Since denominators are equal, add numerators",
                sol("Numerator sum", num1.expr + num2.expr),
                sol("Simplified", sum_num),
                sol("Answer", sum_num / den_lin.expr)
            ),
            time=75
        )
        
    elseif problem_type == 3
        # Multiply two rational expressions
        # ELO: 800-1000
        root1 = nonzero(-7, 7)
        root2 = nonzero(-7, 7)
        a1 = nonzero(-4, 4)
        a2 = nonzero(-4, 4)
        
        while root1 == root2
            root2 = nonzero(-7, 7)
        end
        
        num1 = expand(a1 * (x - root1))
        den1 = expand(x - root2)
        num2 = expand(a2 * (x - root2))
        den2 = expand(x - root1)
        
        ans_num = a1 * a2
        ans_den = 1
        
        problem(
            question="Multiply and simplify: \$\\frac{$(tex(num1))}{$(tex(den1))} \\cdot \\frac{$(tex(num2))}{$(tex(den2))}\$",
            answer=ans_num,
            difficulty=(800, 1000),
            solution=steps(
                "Multiply numerators and denominators",
                sol("Product", (num1 * num2) / (den1 * den2)),
                "Factor and cancel \$(x - $(root1))\$ and \$(x - $(root2))\$",
                sol("Simplified", ans_num * ans_den),
                sol("Answer", ans_num)
            ),
            time=100
        )
        
    elseif problem_type == 4
        # Divide two rational expressions
        # ELO: 850-1050
        common = nonzero(-6, 6)
        a = nonzero(-5, 5)
        b = nonzero(-8, 8)
        c = nonzero(-8, 8)
        
        num1 = expand(a * (x - common))
        den1 = expand(x - b)
        num2 = expand(x - common)
        den2 = expand(x - c)
        
        ans_num = a * (x - c)
        ans_den = x - b
        
        problem(
            question="Divide and simplify: \$\\frac{$(tex(num1))}{$(tex(den1))} \\div \\frac{$(tex(num2))}{$(tex(den2))}\$",
            answer="$(tex(expand(ans_num)))/$(tex(ans_den))",
            difficulty=(850, 1050),
            solution=steps(
                "Multiply by reciprocal: \$\\frac{$(tex(num1))}{$(tex(den1))} \\cdot \\frac{$(tex(den2))}{$(tex(num2))}\$",
                "Factor and cancel common term \$(x - $(common))\$",
                sol("Numerator", expand(a * (x - c))),
                sol("Denominator", x - b),
                sol("Answer", expand(ans_num) / ans_den)
            ),
            time=110
        )
        
    else
        # Subtract rational expressions with different denominators
        # ELO: 900-1100
        b1 = nonzero(-6, 6)
        b2 = nonzero(-6, 6)
        
        while b1 == b2
            b2 = nonzero(-6, 6)
        end
        
        a1 = nonzero(-5, 5)
        a2 = nonzero(-5, 5)
        
        den1 = x - b1
        den2 = x - b2
        
        num1 = a1
        num2 = a2
        
        lcd = expand((x - b1) * (x - b2))
        
        new_num1 = expand(a1 * (x - b2))
        new_num2 = expand(a2 * (x - b1))
        
        final_num = simplify(new_num1 - new_num2)
        
        problem(
            question="Subtract: \$\\frac{$(a1)}{$(tex(den1))} - \\frac{$(a2)}{$(tex(den2))}\$",
            answer="$(tex(final_num))/$(tex(lcd))",
            difficulty=(900, 1100),
            solution=steps(
                "Find LCD: \$$(tex(lcd))\$",
                "First fraction: \$\\frac{$(tex(new_num1))}{$(tex(lcd))}\$",
                "Second fraction: \$\\frac{$(tex(new_num2))}{$(tex(lcd))}\$",
                "Subtract numerators: \$$(tex(new_num1)) - ($(tex(new_num2)))\$",
                sol("Answer", final_num / lcd)
            ),
            time=120
        )
    end
end